File: lessons/00-workflow-or-agent/typescript/src/specialists.ts
import type {SpecialistDispatcher, SpecialistInvocation} from "./types.js";

export class RecordingSpecialistDispatcher implements SpecialistDispatcher {
  readonly #invocations: SpecialistInvocation[] = [];

  get invocations(): readonly SpecialistInvocation[] {
    return this.#invocations.map((invocation) => structuredClone(invocation));
  }

  async dispatch(invocation: SpecialistInvocation): Promise<void> {
    this.#invocations.push(Object.freeze(structuredClone(invocation)));
  }
}

File: lessons/00-workflow-or-agent/python/src/boundrelay_m0/specialists.py
from copy import deepcopy

from .types import SpecialistInvocation


class RecordingSpecialistDispatcher:
    def __init__(self) -> None:
        self._invocations: list[SpecialistInvocation] = []

    @property
    def invocations(self) -> tuple[SpecialistInvocation, ...]:
        return tuple(deepcopy(self._invocations))

    def dispatch(self, invocation: SpecialistInvocation) -> None:
        self._invocations.append(deepcopy(invocation))
